src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (size == 0) {
+            head = tail = newNode;
+        } else {
+            linkToTheEnd(newNode);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        Node<T> newNode = new Node<>(value, null, null);
+        if (index == 0) {
+            if (size == 0) {
+                head = tail = newNode;
+            } else {
+                head.prev = newNode;
+                newNode.next = head;
+                head = newNode;
+            }
+        } else if (index == size) {
+            linkToTheEnd(newNode);
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            Node<T> next = prev.next;
+            prev.next = newNode;
+            newNode.prev = prev;
+            newNode.next = next;
+            next.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T valueToReturn;
+        if (index == 0) {
+            valueToReturn = head.value;
+            head = head.next;
+        } else if (index == size - 1) {
+            valueToReturn = tail.value;
+            tail = tail.prev;
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            valueToReturn = prev.next.value;
+            Node<T> next = prev.next.next;
+            prev.next = next;
+            next.prev = prev;
+        }
+        size--;
+        return valueToReturn;
```

Comment:
```suggestion
        Node<T> target = getNode(index);
        unLink(target);
        return target.value;
```
![image](https://github.com/mate-academy/jv-linked-list/assets/20326858/686225c4-d2de-48c9-842a-9e143100d0c4)